skiko/src/awtMain/objectiveC/macos/MetalDevice.mm

```diff
@@ -0,0 +1,150 @@
+#import "MetalDevice.h"
+#import <stdatomic.h>
+
+static CVReturn MetalDeviceDisplayLinkCallback(CVDisplayLinkRef displayLink, const CVTimeStamp *now, const CVTimeStamp *outputTime, CVOptionFlags flagsIn, CVOptionFlags *flagsOut, void *displayLinkContext) {
+    MetalDevice *device = (__bridge MetalDevice *)displayLinkContext;
+
+    [device handleDisplayLinkFired];
+
+    return kCVReturnSuccess;
+}
+
+@implementation MetalDevice {
+    NSScreen *_displayLinkScreen;
+    CVDisplayLinkRef _displayLink;
+
+    dispatch_semaphore_t _presentingBuffersExhaustionSemaphore;
+    NSConditionLock *_vsyncConditionLock;
+    volatile atomic_bool _displayLinkOk;
```

Comment:
Can we move most of the logic to Kotlin, and only keep simply code blocks in Objective C? Specially, sync primitive it is better to write in Kotlin.

Objective C is more complex to read and understand, and it will be harder to maintain, because there is a fewer Objective C developers.

Reply:
Converted code to Kotlin callback-based.

Reply:
Looks more understable now, thanks! There is still a need in improvement. See my other comment about the complex call chain